File: advent_of_code/2024/day1/puzzle1.py
import csv


def find_total_distance(csv_path: str) -> int:
    total = 0
    left, right = [], []

    with open(csv_path) as file:
        input_csv = csv.DictReader(file)
        for line in input_csv:
            left.append(line['left'])
            right.append(line['right'])

    left.sort(reverse=True)
    right.sort(reverse=True)
    while left and right:
        curr_left = int(left.pop())
        curr_right = int(right.pop())
        total += abs(curr_left - curr_right)

    return total


if __name__ == '__main__':
    res = find_total_distance("day1/input.csv")
    print("total distace is:", res)

File: advent_of_code/2024/day1/puzzle1.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"math"
	"os"
	"sort"
	"strconv"
)

func find_total_distance(data [][]string) int {
	res := 0.0
	left := make([]int, 0)
	right := make([]int, 0)
	l := 0
	r := 0
	for _, line := range data {

		l, _ = strconv.Atoi(line[0])
		r, _ = strconv.Atoi(line[1])
		left = append(left, l)
		right = append(right, r)
	}
	sort.Ints(left)
	sort.Ints(right)

	for i := 0; i < len(left); i++ {
		l = left[i]
		r = right[i]
		res += math.Abs(float64(l - r))
	}

	return int(res)
}

func main() {
	// open file
	f, err := os.Open("day1/input.csv")
	if err != nil {
		log.Fatal(err)
	}

	// remember to close the file at the end of the program
	defer f.Close()

	// read csv values using csv.Reader
	csvReader := csv.NewReader(f)
	if _, err := csvReader.Read(); err != nil {
		panic(err)
	}
	data, err := csvReader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}

	// convert records to array of structs
	res := find_total_distance(data)

	// print the array
	fmt.Printf("%+v\n", res)
}
